File: 2. Structural Design Patterns/Flyweight/Javascript/FlyweightTemplate.js
// Flyweight Pattern – JavaScript Template

class Flyweight {
  // intrinsic, immutable data
}

class FlyweightFactory {
  static cache = new Map();
  static get(key){
    if(!this.cache.has(key)) this.cache.set(key, new Flyweight());
    return this.cache.get(key);
  }
}

class Context {
  constructor(key){ this.fw = FlyweightFactory.get(key); }
  operation(/* extrinsic params */){ /* use this.fw + extrinsic */ }
}

(function main(){ new Context('key').operation(); })();

File: 2. Structural Design Patterns/Flyweight/C#/FlyweightTemplate.cs
// Flyweight Pattern – C# Template

using System.Collections.Generic;

namespace FlyweightTemplate
{
    class Flyweight { /* intrinsic, immutable data */ }

    static class FlyweightFactory
    {
        private static readonly Dictionary<string, Flyweight> cache = new();
        public static Flyweight Get(string key)
        {
            if (!cache.TryGetValue(key, out var fw))
            {
                fw = new Flyweight();
                cache[key] = fw;
            }
            return fw;
        }
    }

    class Context
    {
        private readonly Flyweight _fw;
        public Context(string key) { _fw = FlyweightFactory.Get(key); }
        public void Operation(/* extrinsic params */) { /* use _fw + extrinsic */ }
    }

    class Program { static void Main(){ var c = new Context("key"); c.Operation(); } }
}
